drivers/led/lp5562.c

```diff
@@ -0,0 +1,837 @@
+#define LP5562_PROG_MEM_ENG3_BASE 0x50
+#define LP5562_LED_MAP            0x70
+
+#define LP5562_MIN_BLINK_PERIOD K_MSEC(1)
+#define LP5562_MAX_BLINK_PERIOD K_MSEC(1000)
+
+/* Brightness limits in percent */
+#define LP5562_MIN_BRIGHTNESS 0
+#define LP5562_MAX_BRIGHTNESS 100
+
+/* Values for ENABLE register. */
+#define LP5562_ENABLE_CHIP_EN (1 << 6)
+#define LP5562_ENABLE_LOG_EN  (1 << 7)
+
+/* Values for CONFIG register. */
+#define LP5562_CONFIG_EXTERNAL_CLOCK         0x00
+#define LP5562_CONFIG_INTERNAL_CLOCK         0x01
+#define LP5562_CONFIG_CLOCK_AUTOMATIC_SELECT 0x02
+#define LP5562_CONFIG_PWRSAVE_EN             (1 << 5)
+/* Enable 558 Hz frequency for PWM. Default is 256. */
+#define LP5562_CONFIG_PWM_HW_FREQ_558        (1 << 6)
+
+/* Values for execution engine programs. */
+#define LP5562_PROG_COMMAND_SET_PWM (1 << 6)
+#define LP5562_PROG_COMMAND_RAMP_TIME(prescale, step_time) \
+	(((prescale) << 6) | (step_time))
+#define LP5562_PROG_COMMAND_STEP_COUNT(fade_direction, count) \
+	(((fade_direction) << 7) | (count))
+
+/* Helper definitions. */
+#define LP5562_PROG_MAX_COMMANDS 16
+#define LP5562_MASK              0x03
+#define LP5562_CHANNEL_MASK(channel) ((LP5562_MASK) << (channel << 1))
+
+/*
+ * Available channels. There are four LED channels usable with the LP5562. While
+ * they can be mapped to LEDs of any color, the driver's typical application is
+ * with a red, a green, a blue and a white LED. Since the data sheet's
+ * nomenclature uses RGBW, we keep it that way.
+ */
+enum lp5562_led_channels {
+	LP5562_CHANNEL_B,
+	LP5562_CHANNEL_G,
+	LP5562_CHANNEL_R,
+	LP5562_CHANNEL_W,
+
+	LP5562_CHANNEL_COUNT,
+};
+
+/*
+ * Each channel can be driven by directly assigning a value between 0 and 255 to
+ * it to drive the PWM or by one of the three execution engines that can be
+ * programmed for custom lighting patterns in order to reduce the I2C traffic
+ * for repetitive patterns.
+ */
+enum lp5562_led_sources {
+	LP5562_SOURCE_PWM,
+	LP5562_SOURCE_ENGINE_1,
+	LP5562_SOURCE_ENGINE_2,
+	LP5562_SOURCE_ENGINE_3,
+
+	LP5562_SOURCE_COUNT,
+};
+
+/* Operational modes of the execution engines. */
+enum lp5562_engine_op_modes {
+	LP5562_OP_MODE_DISABLED = 0x00,
+	LP5562_OP_MODE_LOAD = 0x01,
+	LP5562_OP_MODE_RUN = 0x02,
+	LP5562_OP_MODE_DIRECT_CTRL = 0x03,
+};
+
+/* Execution state of the engines. */
+enum lp5562_engine_exec_states {
+	LP5562_ENGINE_MODE_HOLD = 0x00,
+	LP5562_ENGINE_MODE_STEP = 0x01,
+	LP5562_ENGINE_MODE_RUN = 0x02,
+	LP5562_ENGINE_MODE_EXEC = 0x03,
+};
+
+/* Fading directions for programs executed by the engines. */
+enum lp5562_engine_fade_dirs {
+	LP5562_FADE_UP = 0x00,
+	LP5562_FADE_DOWN = 0x01,
+};
+
+struct lp5562_data {
+	struct device *i2c;
+	struct led_data dev_data;
+};
+
+/*
+ * @brief Get the register for the given LED channel used to directly write a
+ *	brightness value instead of using the execution engines.
+ *
+ * @param channel LED channel.
+ * @param source Buffer for the register.
```

Comment:
How about
@param reg Pointer to the register address

Reply:
ack